File: ios/PaperlessTools/Services/TextDiffService.swift
import Foundation

struct DiffPart: Identifiable {
    enum Kind {
        case unchanged
        case added
        case removed
    }

    let id = UUID()
    let kind: Kind
    let text: String
}

struct DiffStats {
    let added: Int
    let removed: Int
    let similarity: Int
}

enum TextDiffService {
    static func diffWords(original: String, updated: String) -> (left: [DiffPart], right: [DiffPart], stats: DiffStats) {
        let oldTokens = tokenize(original)
        let newTokens = tokenize(updated)
        let lcs = longestCommonSubsequence(oldTokens, newTokens)

        var left: [DiffPart] = []
        var right: [DiffPart] = []
        var oldIndex = 0
        var newIndex = 0
        var lcsIndex = 0

        while oldIndex < oldTokens.count || newIndex < newTokens.count {
            if lcsIndex < lcs.count, oldIndex < oldTokens.count, oldTokens[oldIndex] == lcs[lcsIndex] {
                appendPart(&left, kind: .unchanged, token: oldTokens[oldIndex])
                appendPart(&right, kind: .unchanged, token: newTokens[newIndex])
                oldIndex += 1
                newIndex += 1
                lcsIndex += 1
            } else if newIndex < newTokens.count, (lcsIndex >= lcs.count || newTokens[newIndex] != lcs[lcsIndex]) {
                appendPart(&right, kind: .added, token: newTokens[newIndex])
                newIndex += 1
            } else if oldIndex < oldTokens.count {
                appendPart(&left, kind: .removed, token: oldTokens[oldIndex])
                oldIndex += 1
            }
        }

        return (left, right, stats(left: left, right: right))
    }

    private static func tokenize(_ text: String) -> [String] {
        guard !text.isEmpty else { return [] }
        var tokens: [String] = []
        var current = ""

        for character in text {
            if character.isWhitespace {
                if !current.isEmpty {
                    tokens.append(current)
                    current = ""
                }
                tokens.append(String(character))
            } else {
                current.append(character)
            }
        }

        if !current.isEmpty {
            tokens.append(current)
        }

        return tokens
    }

    private static func appendPart(_ parts: inout [DiffPart], kind: DiffPart.Kind, token: String) {
        if let last = parts.last, last.kind == kind {
            parts[parts.count - 1] = DiffPart(kind: kind, text: last.text + token)
        } else {
            parts.append(DiffPart(kind: kind, text: token))
        }
    }

    private static func longestCommonSubsequence(_ a: [String], _ b: [String]) -> [String] {
        let m = a.count
        let n = b.count
        var dp = Array(repeating: Array(repeating: 0, count: n + 1), count: m + 1)

        for i in 1...m {
            for j in 1...n {
                if a[i - 1] == b[j - 1] {
                    dp[i][j] = dp[i - 1][j - 1] + 1
                } else {
                    dp[i][j] = max(dp[i - 1][j], dp[i][j - 1])
                }
            }
        }

        var result: [String] = []
        var i = m
        var j = n
        while i > 0, j > 0 {
            if a[i - 1] == b[j - 1] {
                result.append(a[i - 1])
                i -= 1
                j -= 1
            } else if dp[i - 1][j] >= dp[i][j - 1] {
                i -= 1
            } else {
                j -= 1
            }
        }

        return result.reversed()
    }

    private static func stats(left: [DiffPart], right: [DiffPart]) -> DiffStats {
        var added = 0
        var removed = 0
        var unchanged = 0

        for part in left {
            let count = wordCount(part.text)
            switch part.kind {
            case .added: added += count
            case .removed: removed += count
            case .unchanged: unchanged += count
            }
        }

        for part in right where part.kind == .added {
            added += wordCount(part.text)
        }

        let total = added + removed + unchanged
        let similarity = total > 0 ? Int((Double(unchanged) / Double(total)) * 100) : 100
        return DiffStats(added: added, removed: removed, similarity: similarity)
    }

    private static func wordCount(_ text: String) -> Int {
        text.split { $0.isWhitespace }.filter { !$0.isEmpty }.count
    }
}

File: android/PaperlessTools/app/src/main/kotlin/tools/paperless/app/services/TextDiffService.kt
package tools.paperless.app.services

enum class DiffKind {
    UNCHANGED,
    ADDED,
    REMOVED,
}

data class DiffPart(
    val kind: DiffKind,
    val text: String,
)

data class DiffStats(
    val added: Int,
    val removed: Int,
    val similarity: Int,
)

object TextDiffService {
    fun diffWords(original: String, updated: String): Triple<List<DiffPart>, List<DiffPart>, DiffStats> {
        val oldTokens = tokenize(original)
        val newTokens = tokenize(updated)
        val lcs = longestCommonSubsequence(oldTokens, newTokens)

        val left = mutableListOf<DiffPart>()
        val right = mutableListOf<DiffPart>()
        var oldIndex = 0
        var newIndex = 0
        var lcsIndex = 0

        while (oldIndex < oldTokens.size || newIndex < newTokens.size) {
            when {
                lcsIndex < lcs.size && oldIndex < oldTokens.size && oldTokens[oldIndex] == lcs[lcsIndex] -> {
                    appendPart(left, DiffKind.UNCHANGED, oldTokens[oldIndex])
                    appendPart(right, DiffKind.UNCHANGED, newTokens[newIndex])
                    oldIndex++
                    newIndex++
                    lcsIndex++
                }
                newIndex < newTokens.size && (lcsIndex >= lcs.size || newTokens[newIndex] != lcs[lcsIndex]) -> {
                    appendPart(right, DiffKind.ADDED, newTokens[newIndex])
                    newIndex++
                }
                oldIndex < oldTokens.size -> {
                    appendPart(left, DiffKind.REMOVED, oldTokens[oldIndex])
                    oldIndex++
                }
            }
        }

        return Triple(left, right, stats(left, right))
    }

    private fun tokenize(text: String): List<String> {
        if (text.isEmpty()) return emptyList()
        val tokens = mutableListOf<String>()
        val current = StringBuilder()

        for (character in text) {
            if (character.isWhitespace()) {
                if (current.isNotEmpty()) {
                    tokens.add(current.toString())
                    current.clear()
                }
                tokens.add(character.toString())
            } else {
                current.append(character)
            }
        }

        if (current.isNotEmpty()) {
            tokens.add(current.toString())
        }

        return tokens
    }

    private fun appendPart(parts: MutableList<DiffPart>, kind: DiffKind, token: String) {
        val last = parts.lastOrNull()
        if (last != null && last.kind == kind) {
            parts[parts.lastIndex] = DiffPart(kind, last.text + token)
        } else {
            parts.add(DiffPart(kind, token))
        }
    }

    private fun longestCommonSubsequence(a: List<String>, b: List<String>): List<String> {
        val m = a.size
        val n = b.size
        val dp = Array(m + 1) { IntArray(n + 1) }

        for (i in 1..m) {
            for (j in 1..n) {
                dp[i][j] = if (a[i - 1] == b[j - 1]) {
                    dp[i - 1][j - 1] + 1
                } else {
                    maxOf(dp[i - 1][j], dp[i][j - 1])
                }
            }
        }

        val result = mutableListOf<String>()
        var i = m
        var j = n
        while (i > 0 && j > 0) {
            when {
                a[i - 1] == b[j - 1] -> {
                    result.add(a[i - 1])
                    i--
                    j--
                }
                dp[i - 1][j] >= dp[i][j - 1] -> i--
                else -> j--
            }
        }

        return result.reversed()
    }

    private fun stats(left: List<DiffPart>, right: List<DiffPart>): DiffStats {
        var added = 0
        var removed = 0
        var unchanged = 0

        for (part in left) {
            val count = wordCount(part.text)
            when (part.kind) {
                DiffKind.ADDED -> added += count
                DiffKind.REMOVED -> removed += count
                DiffKind.UNCHANGED -> unchanged += count
            }
        }

        for (part in right) {
            if (part.kind == DiffKind.ADDED) {
                added += wordCount(part.text)
            }
        }

        val total = added + removed + unchanged
        val similarity = if (total > 0) ((unchanged.toDouble() / total) * 100).toInt() else 100
        return DiffStats(added, removed, similarity)
    }

    private fun wordCount(text: String): Int =
        text.split(Regex("\\s+")).count { it.isNotEmpty() }
}
